p2p/server/server.go

```diff
@@ -195,6 +202,19 @@ func New(h Host, proto string, handler StreamHandler, opts ...Opt) *Server {
 		}
 	}
 
+	srv.limit = rate.NewLimiter(
+		rate.Every(srv.interval/time.Duration(srv.requestsPerInterval)),
+		srv.requestsPerInterval,
+	)
+	srv.sem = semaphore.NewWeighted(int64(srv.queueSize))
+	if srv.metrics != nil {
+		srv.metrics.targetQueue.Set(float64(srv.queueSize))
+		srv.metrics.targetRps.Set(float64(srv.limit.Limit()))
+	}
+	srv.h.SetStreamHandler(protocol.ID(srv.protocol), func(stream network.Stream) {
+		srv.queue <- request{stream: stream, received: time.Now()}
```

Comment:
There was a `select` case in the block that was moved. Any particular reason why to remove the `default` case? Ideally there would be a context here that could be selected on, but also the problem is the rate limiting which could result in messages being dropped (but also, goroutines could pile up waiting to write to this channel, which would have other repercussions, like messages timing out on the other end). It might be useful to dwell on this a bit longer.

A bit of food for thought that maybe doing rate-limiting on the p2p level is not a good approach - some protocols might be more aggressive than others on resources, which could result in resource starvation on others. This means that having rate-limiting on the p2p level won't solve the problem of resources necessarily and also removes the possibility of prioritizing some protocols over others.

Reply:
Check the reading side of the channel - every incoming request is immediately handled and if it exceeds the queue size it will be dropped immediately. I did this so that `TryAcquire` and `Release` calls to the semaphore are in the same code block.

I can add the `TryAcquire` to the callback of `SetStreamHandler` and only if it passes that check send the request to the queue to be handled in `Run` but then the `Release` would be disconnected from the `TryAcquire` call 🤷

Reply:
If the reader goes away then the goroutines pile up on shutdown. I'm not sure which repercussions this may have (or not have). As a rule of thumb I think we should try to avoid writing directly to channels (without `select`s). But that's just me...
It could be that it is inconsequential, but it could potentially have some side-effects. I'm not so familiar with the code.

Reply:
I updated the `SetStreamHandler` callback in a way which does the handling on the sender side of the channel instead of the receiver side, see [b0b6cf9](https://github.com/spacemeshos/go-spacemesh/pull/6097/commits/b0b6cf9fd02aebf87800e09d062f6468d65fd94f). It has the advantage of dropping requests on the sending side of the channel if a) the context is already done or b) the semaphore is full, but since now the callback has to be set in the `Run` method I need an additional check with the `Ready` method because otherwise the first request might be sent from the client to the server before the `SetStreamHandler` callback has been set causing the test to halt.

So it is either:
- waiting on `Ready` after calling `Run`
- handling limiting and context checking in the `Run` method instead of in the `SetStreamHandler` callback

Reply:
OK nvm, I found a way to do it without the need for the `Ready` method 🙂